### PR TITLE
fix: angular 5 element life-cycle not triggered correctly due to pipe

### PR DESCRIPTION
**Pull request by stephenlautier**
We are currently updating to angular 5, and making use of this library.
After the update, we have some components broken, and after some debugging I've noticed it was being caused by the `localize` pipe of this library, its caused with the `ref.detectChanges`, still, the reason why its still bit weird (might be an issue with angular itself tbh).

### Issue
Basically, we have a super simple button directive (simply adds some classes on an element) on init, and once using the directive in conjunction with the localize pipe it's causing the `ngOnInit` to never be invoked.

This PR fixes that and also some other refactoring/improvements.

P.S. I cannot get tests to work I'm getting q the following error:
```
HeadlessChrome 0.0.0 (Windows 10 0.0.0) ERROR
  Uncaught ReferenceError: core_2 is not defined
  at webpack:///src/localize-router.parser.ts:2:0 <- config/spec-bundle.js:23565
```

This is one of the last few issues which are blocking us from updating to angular 5, would highly appreciate if you can look into it a bit urgently. Thanks.

using `localize-router ` version `1.0.0-rc.3`

### CODE REVIEW

File: src/localize-router.pipe.ts [hunk 1]
```diff
@@ -37,7 +34,7 @@ export class LocalizeRouterPipe implements PipeTransform {
     if (!query || query.length === 0 || !this.localize.parser.currentLang) {
       return query;
     }
-    if (equals(query, this.lastKey) && equals(this.lastLanguage, this.localize.parser.currentLang)) {
+    if (!this.markForTransform && query === this.lastKey && this.lastLanguage === this.localize.parser.currentLang) {
```
Comment (stephenlautier) [inline src/localize-router.pipe.ts]: the reason why I remove the `equals` was that a ref compare will be way faster, especially if its invoked too many times, and if the reference changes then simply might as well recompute it

**Comment by stephenlautier** [comment]
Getting same tests error on CI as well

**Comment by meeroslav** [comment]
Hi @stephenlautier , `detectChanges` was used intentionally to fix problems with delayed change detection of `RouterLinkWithHref` directive causing `RouterLinkActive` directive to end up in invalid state after language switch.

I would need to investigate whether your solution didn't revert this fix.

Forcing change detection again has some unplanned side effects and you might just found one of them, so the whole fix might have to be reverted.

Please use version 1.0.0-rc.2 until I review your changes and investigate the problem with build on circleCI.

**Comment by stephenlautier** [comment]
@meeroslav thanks that what I did just now and its working for me for the time being thanks. 
Re `RouterLinkActive` not entirely sure, if you have an easy repro sample I can give it a look.

One thing I can think of if its a solution you might try is to put the `routeLinkActive` on its parent not directly on the `routeLink`, as following:

```html
<li routeLinkActive>
    <a [routeLink]="...">link</a>
</li>
```

after I looked a bit into their code and it refresh on its descendants or when the user navigates. In general, it gives you better flexibility for styling when done this way as well.

**Comment by meeroslav** [comment]
@stephenlautier You can use small demo app in`demo\cli` folder. Just link your build with node_modules in the cli project and you should be able to test it.

As for the build, the master is now green (no injector `core_2` errors) so you can now focus on fixing the remaining ones caused by your changes.

**Comment by meeroslav** [comment]
@stephenlautier In case you haven't use `npm link` before, you can find info [here](https://docs.npmjs.com/cli/link).

To build and bundle the source locally you can use: 
```
npm run prepublish
```

so the whole step by step for testing would look like this:
```
npm run prepublish
npm link
cd demo\cli
npm i
npm link localize-router
ng serve 
```

**Comment by stephenlautier** [comment]
Thanks for the testing tips, linking for some reason im getting the following error:

2017-11-29 23_41_28-package json - localize-router - visual studio code

However, i just copy/paste for now. 
I fixed the tests and now I'm testing out within the sample. Currently, it seems to work except the `routerLinkActive` isn't :(

**Comment by stephenlautier** [comment]
by simply from `this._ref.markForCheck();` to `detectChanges` in my implementation it "almost" works, first lang switch it doesnt work, all the rest it works.
Most probably because it fetches the language and it screws it for the first one

**Pull request closed without merging**